### ISSUE TITLE
Inconsistent use of SecArgumentsLimit in Recommended Rules

### ISSUE THREAD

**Issue by studersi** [SpiderLabs/ModSecurity#3261]
**Describe the bug**

The limit `SecArgumentsLimit` is inconsistently used an documented.

| | v2.9.8 | v3.0.13|
|--|:--:|:--:|
| Supported (code) | [x](https://github.com/owasp-modsecurity/ModSecurity/blob/v2.9.8/apache2/apache2_config.c#L2131-L2153) | ? (could not find it in source code, only in recommended rules) |
| Supported (documentation) | [x (2.9.7)](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-%28v2.x%29#secargumentslimit) | [x (3.0.5)](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-%28v3.x%29#user-content-SecArgumentsLimit) |
| ModSecurity Recommended  Rules| [-](https://github.com/owasp-modsecurity/ModSecurity/blob/v2.9.8/modsecurity.conf-recommended) | [x](https://github.com/owasp-modsecurity/ModSecurity/blob/v3.0.13/modsecurity.conf-recommended#L60-L68) |

**Logs and dumps**

Not applicable.

**To Reproduce**

Not applicable.

**Expected behavior**

I would expect the recommended rules for v2.9.8 to also include the SecArgumentsLimit configuration, like 3.0.13.

**Server (please complete the following information):**
 - ModSecurity version (and connector): v2.9.8
 - WebServer: Not applicable.
 - OS (and distro): Not applicable.

**Rule Set (please complete the following information):**
 - Running any public or commercial rule set? ModSecurity Recommended Rules
 - What is the version number? v2.9.8

**Additional context**

-

**Comment by airween** [comment]
Hi @studersi,

thanks for reporting this. You're completely right, it would be much better to make the both versions' default configuration files consistent.

For the record: in libmodsecurity3 the initial set step of that variable is [here](https://github.com/owasp-modsecurity/ModSecurity/blob/v3/master/src/parser/seclang-parser.cc#L3013-L3017). (This is a "compiled" code (by Bison), the original part is [here](https://github.com/owasp-modsecurity/ModSecurity/blob/v3/master/src/parser/seclang-parser.yy#L1594-L1598)).

Would you mind to send a PR for v2 to fix this?

My side note: I'm not sure [this](https://github.com/owasp-modsecurity/ModSecurity/blob/v3.0.13/modsecurity.conf-recommended#L67-L68) rule makes sense; if the engine is in `DetectionOnly` mode (which is the default...!) then this rule does nothing... But of course we can start to discuss, what would be the nice solution.

**Comment by dune73** [comment]
I'm not sure I get your argument about the engine. All the blocking recommended rules do nothing unless you put the engine in - well - blocking mode. So what's the difference here?

**Comment by airween** [comment]
There is no difference, I just think it's a bit confuse that in the "recommended" configuration file we set the engine to "DetectionOnly" and add a rule without any notification that won't work after installation.

My other comment regarding this rule that rule [200002](https://github.com/owasp-modsecurity/ModSecurity/blob/v3.0.13/modsecurity.conf-recommended#L75-L76) does (almost) the same (but in another way). If the `SecArgumentsLimit` is set (and it is [here](https://github.com/owasp-modsecurity/ModSecurity/blob/v3.0.13/modsecurity.conf-recommended#L62)), then if the number of arguments has reached the limit, then it sets the `REQBODY_ERROR`, which triggers rule `200002`.

I really don't know that rule `200007` is necessary. But not really disturbing, just my 2 cents.

**Comment by airween** [comment]
@studersi, @dune73, @marcstern - what do you think guys, do we need rule 200007 in recommended `modsecurity.conf` file? As I explained above I don't think it's necessary, but I don't want to decide alone. I will gladly create a PR if it's really need.

**Comment by dune73** [comment]
I see the redundancy and keeping multiple values in sync is annoying.

If we recommend 200007, then I think the comment for `SecArgumentsLimit` should make it clear that a violation will lead to a `REQBODY_ERROR` and thus 200002 being triggered.

Also, I am not sure about the implications of the comment `If SecArgumentsLimit has been set, you probably want to reject any request body that has only been partly parsed.`. What does this really mean?

And is there any influence of the different body parsers on reqbody error and the argument limit behavior?

**Comment by marcstern** [comment]
SecArgumentsLimit + 200002 is better than 200007 because the former will stop the JSON parsing after X args (and JSON parsing is CPU-intensive). For other parsers, there's no difference (currently).
200007 is totally redundant, but less powerful. We should remove it.

**Comment by airween** [comment]
> And is there any influence of the different body parsers on reqbody error and the argument limit behavior?

Yes, for eg. JSON parser will stop too if it reaches the arglimit. (And the planned XML parser (under #3178) will follow this behavior too.)

@marcstern is right, rule `200007` would wait until the engine counts the number of arguments (which never occurs because rule `200002` prevents that).

**Comment by studersi** [comment]
I agree that `200002` is probably sufficient for most cases and `200007` is not strictly necessary when it comes to blocking the request.

I wonder, whether this was added to log additional context for these requests as to why the request was blocked. If I remember correctly `SecArgumentsLimit` itself does not log any relevant information and you only get a generic body parsing error in the logs with no mention of `SecArgumentsLimit`.

Having a dedicated error message might be very helpful when debugging these issues.

**Comment by dune73** [comment]
That's a good thought. Would make sense.

**Comment by marcstern** [comment]
The REQBODY_ERROR is populated with the right error message ("SecArgumentsLimit exceeded" or "More than %ld JSON keys"). This variable should always be logged in case it's populated.

**Comment by studersi** [comment]
In that case, I don't think I see any reason to keep `200007` either.
